# KNN

In [1]:
%load_ext autoreload
%autoreload 2
from preprocessing import *
from sklearn.model_selection import train_test_split, GridSearchCV, StratifiedKFold, cross_val_score
from sklearn.pipeline import Pipeline
from sklearn.neighbors import KNeighborsClassifier
from sklearn import metrics
import numpy as np
import pandas as pd
from joblib import dump, load

In [2]:
SCORINGS = ["roc_auc", "accuracy", "precision", "recall", "f1"]
METRIC = "roc_auc"

def grid_table(grid):
    tabla = pd.DataFrame(grid.cv_results_)
    tabla.sort_values("rank_test_" + METRIC, inplace = True)
    tabla.reset_index(inplace = True)
    cols = ["param_n_neighbors"] + ["mean_test_" + x for x in SCORINGS]
    return tabla[cols]

def metrics_table(model, X, Y):
    predicted = model.predict(X)
    probabilities = model.predict_proba(X)[:,1]
    
    roc_auc = metrics.roc_auc_score(Y, probabilities)
    f1 = metrics.f1_score(Y, predicted)
    acc = metrics.accuracy_score(Y, predicted)
    rec = metrics.recall_score(Y, predicted)
    prec = metrics.precision_score(Y, predicted)
    return pd.DataFrame.from_dict({
        "AUC-ROC": [roc_auc], "Accuracy": [acc], "Precision": [prec], "Recall": [rec], "F1 Score": [f1]
    })

## Modelo con pesos uniformes

Vamos a hacer un Grid Search para encontrar el valor óptimo de vecinos a utilizar.

In [17]:
initialize_dataset()

df_features = pd.read_csv("datasets/df_features.csv", low_memory = False, index_col = "id")
df_target = pd.read_csv("datasets/df_target.csv", low_memory = False, index_col = "id")

common(df_features, df_target)
viento_trigonometrico(df_features)
df_features = hashing_trick(df_features, 24, "barrio")
pipe = iterative_imputer()
pipe = standarizer(pipe)

X_train, X_test, Y_train, Y_test = train_test_split(df_features, df_target, test_size = 0.20, random_state = 123)

In [4]:
# La grid tarda muchísimo en terminar, si se desea se puede cargar de archivo y saltear el próximo bloque
# de código, descomentando la siguiente línea. También se pueden saltear los próximos 3 bloques de código si
# solo se desea ver el modelo final con los parámetros encontrados.
# grid = load('KNN/uniforme_grid.joblib')

In [5]:
X_grid = pipe.fit_transform(X_train)

# Cuidado ejecutando esto que ya con 6 threads consume como 13GB de RAM a momentos y puede morir la compu,
# cualquier cosa bajar el n_jobs a un número más chico (o más grande si tenes +16GB y va a ir más rápido)
grid = GridSearchCV(KNeighborsClassifier(weights='uniform'), param_grid = {'n_neighbors': range(1, 76, 2)},
                    verbose = 1, n_jobs = 6, cv = StratifiedKFold(3), scoring = SCORINGS, refit = METRIC)

grid.fit(X_grid, Y_train.values.ravel())

grid.best_score_

Fitting 3 folds for each of 38 candidates, totalling 114 fits


0.8694025603157335

In [6]:
dump(grid, 'KNN/uniforme_grid.joblib') 

['KNN/uniforme_grid.joblib']

In [12]:
grid_table(grid)

param_n_neighbors  mean_test_roc_auc  mean_test_accuracy  \
0                 55           0.869403            0.835312   
1                 47           0.869398            0.836912   
2                 45           0.869295            0.837169   
3                 49           0.869283            0.836546   
4                 53           0.869269            0.835801   
5                 51           0.869208            0.836192   
6                 59           0.869189            0.835434   
7                 57           0.869138            0.835300   
8                 71           0.869089            0.833187   
9                 67           0.869081            0.833724   
10                61           0.869072            0.834775   
11                65           0.869069            0.834115   
12                73           0.869047            0.832918   
13                69           0.869041            0.833895   
14                63           0.868998            0.834298   
15                75           0.868958            0.832820   
16                43           0.868953            0.837108   
17                41           0.868579            0.837682   
18                39           0.868155            0.838427   
19                35           0.868004            0.839294   
20                37           0.867773            0.838610   
21                33           0.867571            0.839453   
22                31           0.867076            0.839941   
23                29           0.866495            0.839868   
24                27           0.865837            0.840381   
25                25           0.865408            0.840503   
26                23           0.864773            0.841028   
27                21           0.863785            0.840882   
28                19           0.862207            0.840735   
29                17           0.860316            0.840906   
30                15           0.858206            0.841798   
31                13           0.854910            0.841346   
32                11           0.850124            0.841370   
33                 9           0.843808            0.840088   
34                 7           0.833735            0.837804   
35                 5           0.817418            0.834848   
36                 3           0.786423            0.825773   
37                 1           0.690139            0.795603   

    mean_test_precision  mean_test_recall  mean_test_f1  
0              0.794317          0.356121      0.491711  
1              0.790530          0.368789      0.502898  
2              0.789082          0.371519      0.505136  
3              0.791385          0.365840      0.500328  
4              0.794926          0.358578      0.494170  
5              0.792473          0.362837      0.497713  
6              0.800443          0.352299      0.489189  
7              0.797487          0.353664      0.489954  
8              0.805004          0.335700      0.473766  
9              0.803598          0.339849      0.477642  
10             0.800411          0.348368      0.485399  
11             0.802925          0.342689      0.480302  
12             0.806438          0.333133      0.471462  
13             0.805863          0.339303      0.477497  
14             0.801505          0.344764      0.482090  
15             0.807799          0.331604      0.470166  
16             0.785242          0.374249      0.506878  
17             0.785906          0.377252      0.509762  
18             0.786765          0.381075      0.513425  
19             0.784844          0.388009      0.519277  
20             0.784020          0.384515      0.515953  
21             0.781387          0.392050      0.522104  
22             0.779073          0.397237      0.526138  
23             0.775829          0.399803      0.527640  
24             0.776605          0.402261      0.529968  
25             0.772989          0.40

La progresión de los parámetros es interesante en este caso. Podemos ver como los valores más altos de Recall son para cantidad de vecinos más pequeñas, ya que si bien es más sensible a outliers permite a su vez detectar más fácilmente los casos positivos al tener el dataset tiene muchos más casos negativos. Este desbalance provoca que al aumentar la cantidad de vecinos a utilizar aumente las probabilidades de que prediga como negativo, dando como resultado peor recall pero más precision.

Si bien el Recall es pésimo y hay otros parámetros con mejor accuracy, tomando como criterio el ROC-AUC como dice la consigna, el mejor modelo resultó ser el de K=55. Vamos a probar un modelo con todos los datos y ese parámetro:

In [18]:
pipe.steps.append(['nb', KNeighborsClassifier(n_neighbors = 55, weights='uniform')])

pipe.fit(X_train, Y_train.values.ravel())

metrics_table(pipe, X_test, Y_test.values.ravel())

AUC-ROC  Accuracy  Precision    Recall  F1 Score
0  0.873361  0.837551   0.804983  0.365376  0.502618

Validando con el set de test, el AUC-ROC dio 0.873 (incluso un poco mejor que en el grid) y el accuracy 83,8%. El Recall dio exageradamente bajo por lo explicado previamente.

## Ponderados según la distancia

Vamos a hacer el mismo procedimiento pero con pesos variables tal que la predicción le de más importancia a las instancias más cercanas, a ver si mejora. Como afecta significativamente el modelo, vamos a repetir el Grid Search por si resulta mejor otro valor.

In [7]:
initialize_dataset()

df_features = pd.read_csv("datasets/df_features.csv", low_memory = False, index_col = "id")
df_target = pd.read_csv("datasets/df_target.csv", low_memory = False, index_col = "id")

common(df_features, df_target)
viento_trigonometrico(df_features)
df_features = hashing_trick(df_features, 24, "barrio")
pipe2 = iterative_imputer()
pipe2 = standarizer(pipe2)

X_train, X_test, Y_train, Y_test = train_test_split(df_features, df_target, test_size = 0.20, random_state = 123)

In [8]:
# La grid tarda muchísimo en terminar, si se desea se puede cargar de archivo y saltear el próximo bloque
# de código, descomentando la siguiente línea. También se pueden saltear los próximos 3 bloques de código si
# solo se desea ver el modelo final con los parámetros encontrados.
# grid = load('KNN/ponderado_grid.joblib')

In [9]:
X_grid = pipe2.fit_transform(X_train)

# Cuidado ejecutando esto que ya con 6 threads consume como 13GB de RAM a momentos y puede morir la compu,
# cualquier cosa bajar el n_jobs a un número más chico (o más grande si tenes +16GB y va a ir más rápido)
grid2 = GridSearchCV(KNeighborsClassifier(weights = "distance"), param_grid = {'n_neighbors': range(1, 76, 2)},
                    verbose = 1, n_jobs = 6, cv = StratifiedKFold(3), scoring = SCORINGS, refit = METRIC)

grid2.fit(X_grid, Y_train.values.ravel())

grid2.best_score_

Fitting 3 folds for each of 38 candidates, totalling 114 fits


0.87113215674804

In [10]:
dump(grid2, 'KNN/ponderado_grid.joblib') 

['KNN/ponderado_grid.joblib']

In [11]:
grid_table(grid2)

param_n_neighbors  mean_test_roc_auc  mean_test_accuracy  \
0                 55           0.871132            0.836509   
1                 47           0.871070            0.838146   
2                 53           0.871001            0.836985   
3                 49           0.870979            0.837914   
4                 59           0.870966            0.836448   
5                 45           0.870965            0.838231   
6                 51           0.870924            0.837682   
7                 57           0.870911            0.836448   
8                 71           0.870878            0.834457   
9                 61           0.870875            0.835727   
10                67           0.870869            0.835043   
11                73           0.870856            0.834396   
12                65           0.870853            0.835312   
13                69           0.870841            0.834811   
14                63           0.870804            0.835715   
15                75           0.870785            0.834274   
16                43           0.870649            0.838219   
17                41           0.870309            0.838671   
18                39           0.869916            0.839331   
19                35           0.869717            0.840112   
20                37           0.869569            0.839660   
21                33           0.869254            0.840186   
22                31           0.868777            0.840809   
23                29           0.868230            0.840931   
24                27           0.867566            0.841187   
25                25           0.867114            0.841676   
26                23           0.866411            0.841676   
27                21           0.865412            0.841627   
28                19           0.863847            0.841358   
29                17           0.861994            0.841273   
30                15           0.859866            0.842225   
31                13           0.856619            0.841603   
32                11           0.851797            0.841407   
33                 9           0.845454            0.840076   
34                 7           0.835587            0.837804   
35                 5           0.819351            0.834933   
36                 3           0.787901            0.825736   
37                 1           0.690139            0.795603   

    mean_test_precision  mean_test_recall  mean_test_f1  
0              0.793241          0.364148      0.499103  
1              0.790965          0.375888      0.509561  
2              0.793343          0.366933      0.501745  
3              0.792303          0.373376      0.507528  
4              0.798991          0.359397      0.495724  
5              0.788498          0.378454      0.511398  
6              0.793997          0.370646      0.505328  
7              0.795742          0.361855      0.497424  
8              0.803125          0.344491      0.482124  
9              0.797592          0.356121      0.492347  
10             0.802333          0.348531      0.485932  
11             0.805024          0.342798      0.480810  
12             0.800756          0.351316      0.488322  
13             0.802670          0.346948      0.484439  
14             0.800862          0.353609      0.490562  
15             0.806164          0.341269      0.479504  
16             0.786297          0.380201      0.512531  
17             0.785786          0.383423      0.515341  
18             0.786520          0.386808      0.518552  
19             0.783920          0.393852      0.524280  
20             0.784436          0.390630      0.521520  
21             0.781468          0.396527      0.526078  
22             0.779824          0.401933      0.530423  
23             0.776943          0.405428      0.532768  
24             0.776212          0.407666      0.534543  
25             0.774604          0.41

Los resultados son prácticamente idénticos. Vamos a probar a probar K=55 y distancias ponderadas con el set entero.

In [15]:
pipe2.steps.append(['nb', KNeighborsClassifier(n_neighbors = 55, weights = "distance")])

pipe2.fit(X_train, Y_train.values.ravel())

metrics_table(pipe, X_test, Y_test.values.ravel())

AUC-ROC  Accuracy  Precision    Recall  F1 Score
0  0.873361  0.837551   0.804983  0.365376  0.502618

Dio exactamente los mismos resultados. Parece que ese parámetro no afecta mucho al modelo.